### PR TITLE
Timo Tyrrell BE 1710

### PR DESCRIPTION
**Pull request by timomitchel**

### CODE REVIEW

**Comment by timomitchel** [comment]
Iteration 4 complete

File: lib/activity.rb [hunk 1]
```diff
@@ -1,2 +1,30 @@
...
+
+  attr_reader :name, :participants
+
+  def initialize(name)
+    @name = name
+    @participants = Hash.new
+  end
+
+  def add_participant(name, amount)
+    @participants.store(name, amount)
```
Comment (VictoriaVasys) [inline lib/activity.rb]: Although this works, it's better Ruby style to create key-value pairs with the bracket notation.

File: lib/activity.rb [hunk 2]
```diff
@@ -1,2 +1,30 @@
...
+  end
+
+  def total_cost
+    @participants.values.sum
+  end
+
+  def split_cost
+    total_cost / @participants.values.length
```
Comment (VictoriaVasys) [inline lib/activity.rb]: The #values method is overkill here. What does #length calculate for a hash?

File: lib/activity.rb [hunk 3]
```diff
@@ -1,2 +1,30 @@
...
+  end
+
+  def amount_owed
+    @participants.keys.reduce({}) do |result, name|
+      result.store(name, split_cost - @participants[name])
+      result
+    end
+  end
```
Comment (VictoriaVasys) [inline lib/activity.rb]: See my note about #store above. Also, it's better to iterate over the whole hash, especially when accessing parts of it within your block.

File: lib/reunion.rb [hunk 4]
```diff
@@ -1,2 +1,45 @@
...
+    end.sum
+  end
+
+  def activity_1_amounts
+    activity_totals[0]
+  end
+
+  def activity_2_amounts
+    activity_totals[1]
+  end
```
Comment (VictoriaVasys) [inline lib/reunion.rb]: This method and the one above are not dynamic; you should strive to only implement dynamic methods and use enums to access the pieces of your collection instead.

File: lib/reunion.rb [hunk 5]
```diff
@@ -1,2 +1,45 @@
...
+
+  def activity_totals
+    @activities.map do |activity|
+      activity.amount_owed
+    end
+  end
```
Comment (VictoriaVasys) [inline lib/reunion.rb]: Be a little more clear on naming this method since you have other totals in your program.

File: lib/reunion.rb [hunk 6]
```diff
@@ -1,2 +1,45 @@
...
+
+  def total_owed
+    activity_1_amounts.merge(activity_2_amounts) do |key, first_val, second_val|
+      first_val + second_val
+    end
+  end
```
Comment (VictoriaVasys) [inline lib/reunion.rb]: I think there's a better way to iterate through your activities and calculate the owed amounts. Play around with different enums here steering away from using your activity_#_amounts methods.

File: test/activity_test.rb [hunk 7]
```diff
@@ -2,6 +2,66 @@
...
+
+    assert_equal 'Hike', activity.name
+    assert_equal Hash.new, activity.participants
+  end
+
+  def test_can_add_participant_to_activity
+    activity = Activity.new('Hike')
+    activity.add_participant('Timo', 5)
+
+    assert_equal 'Hike', activity.name
```
Comment (VictoriaVasys) [inline test/activity_test.rb]: This is a redundant assertation since you already asserted it above.

File: test/activity_test.rb [hunk 8]
```diff
@@ -2,6 +2,66 @@
...
+  end
+
+  def test_can_hold_multiple_participants
+    activity = Activity.new('Hike')
+    activity.add_participant('Timo', 5)
+    activity.add_participant( 'Katy', 10)
+    activity.add_participant('Erika', 0)
+
+    assert_equal ["Timo", "Katy", "Erika"], activity.participants.keys
+    assert_equal [5, 10, 0], activity.participants.values
```
Comment (VictoriaVasys) [inline test/activity_test.rb]: It would be more explicit to test against the whole hash.

File: test/reunion_test.rb [hunk 9]
```diff
@@ -4,4 +4,80 @@
...
+  end
+
+  def test_it_can_hold_attributes
+    reunion = Reunion.new('house')
+
+    assert_equal 'house', reunion.location
+    assert_equal [], reunion.activities
+  end
+
+  def test_add_activities
```
Comment (VictoriaVasys) [inline test/reunion_test.rb]: Try to be a little more descriptive with your test names; this is testing `#add_activity`, so a better name might be `test_add_activity_builds_the_activities_collection` or something; always include an explanation of what the method should do.

File: test/reunion_test.rb [hunk 10]
```diff
@@ -4,4 +4,80 @@
...
+    activity_2 = Activity.new('hang')
+    activity_2.add_participant('Timo', 5)
+    activity_2.add_participant( 'Katy', 10)
+    activity_2.add_participant('Erika', 0)
+
+    reunion.add_activity(activity_1)
+    reunion.add_activity(activity_2)
+
+    assert_equal ({"Timo"=>0, "Katy"=>-10, "Erika"=>10}), reunion.total_owed
+  end
```
Comment (VictoriaVasys) [inline test/reunion_test.rb]: It'd be best to start smaller and build toward more complex inputs. You should also have various tests with different numbers of participants, etc. for edge cases.
